src/chain/tezos/TezosNodeWriter.ts

```diff
@@ -641,29 +695,83 @@ export namespace TezosNodeWriter {
+     *
+     * Inspired by OperationFeeEstimator.estimateAndApplyFees from @tacoinfra/harbinger-lib.
+     * 
+     * @see https://github.com/tacoinfra/harbinger-lib/blob/c27cdd4643574c245ddca62bf7013964ae684bd5/src/operation-fee-estimator.ts#L30
+     *
+     * @param {string} server Tezos node to connect to
+     * @param {string} chainid The chain ID to apply the operation on. 
+     * @param operations An array of operations to process.
+     * @returns {[gas: number, storageCost: number], estimatedFee: number, estimatedStorageBurn: number}
+     */
+    export async function estimateOperationGroup(server: string, chainid: string, operations: Array<TezosP2PMessageTypes.StackableOperation>): Promise<any> {
+        let operationResources: { gas: number, storageCost: number }[] = [];
+
+        for (let i = 0; i < operations.length; i++) { // Estimate each operation.
+            const operation = operations[i];
+
+            // Estimate resources used in the set of prior transactions.
+            // If there were no prior transactions, set resource usage to 0.
+            let priorConsumedResources = { gas: 0, storageCost: 0 };
+            if (i > 0) {
+                const priorTransactions = operations.slice(0, i);
+                priorConsumedResources = await estimateOperation(server, chainid, ...priorTransactions);
+            }
+
+            // Estimate resources for everything up to the current transaction. Newer transactions may depend on previous transactions, thus all transactions must be estimated.
+            const currentTransactions = operations.slice(0, i + 1);
+            const currentConsumedResources = await TezosNodeWriter.estimateOperation(server, chainid, ...currentTransactions);
+
+            // Find the actual transaction cost by calculating the delta between the two transactions resource usages.
+            const gasLimitDelta = currentConsumedResources.gas - priorConsumedResources.gas;
+            const storageLimitDelta = currentConsumedResources.storageCost - priorConsumedResources.storageCost;
+
+            operationResources.push(
+                {
+                    gas: gasLimitDelta + TezosConstants.GasLimitPadding,
+                    storageCost: storageLimitDelta + TezosConstants.StorageLimitPadding
+                }
+            );
         }
+
+        const gasLimitTotal = operationResources.map(r => r.gas).reduce((a, c) => a + c, 0);
+        const storageLimitTotal = operationResources.map(r => r.storageCost).reduce((a, c) => a + c, 0);
+        const forgedOperationGroup = forgeOperations('BMLxA4tQjiu1PT2x3dMiijgvMTQo8AVxkPBPpdtM8hCfiyiC1jz', operations);
+        const groupSize = forgedOperationGroup.length / 2 + 64; // operation group bytes + signature bytes
+        let estimatedFee = Math.ceil(gasLimitTotal / 10) + TezosConstants.BaseOperationFee + groupSize + TezosConstants.DefaultBakerVig;
```

Comment:
maybe move these into little helper funcs?

Reply:
will consider it for 508